hal/armv7m/mpu.c

```diff
@@ -0,0 +1,322 @@
+				srdMask |= 1u << bit;
+			else
+				*allocSize += srSize;
+
+			srBase += srSize;
+		}
+	}
+
+	return mpu_regionSet(region, addr & alignMask, alignBits, attr, enable, srdMask);
+}
+
+
+/* Allocate MPU sub/regions for a map */
+static int mpu_regionAlloc(u8 *regCnt, u32 mapId, u32 addr, u32 size, u32 attr, u8 enable)
+{
+	u32 allocSize = 0;
+	u8 regCur = *regCnt;
+
+	while (size > 0 && (regCur - *regCnt) < 2) {
+		if (mpu_regionBestFit(regCur, addr, size, attr, enable, &allocSize) < 0)
+			return -EPERM;
+
+		if (allocSize > size)
+			return -EPERM;
+
+		addr += allocSize;
+		size -= allocSize;
+		regCur++;
+	}
+
+	if (size != 0)
+		return -EPERM;
+
+	for (;;) {
+		mpu_common.hal.mpu.map[*regCnt] = mapId;
+		if (*regCnt >= regCur)
+			break;
+		(*regCnt)++;
+	}
+
+	return EOK;
+}
+
+
+/* Allocate MPU sub/regions for a map */
+int mpu_mapAlloc(addr_t start, addr_t end, u32 attr, u32 id)
+{
+	int res;
+	if ((res = mpu_regionAlloc(&mpu_common.regCnt, id, start, end - start, attr, 1)) < 0) {
+		mpu_common.flags |= mpu_fError;
+		return res;
+	}
+
+	mpu_common.mapsCnt++;
+
+	return EOK;
+}
+
+
+/* Prints allocated MPU regions */
+void mpu_mapInfo(void)
+{
+	unsigned int i;
+	mpu_region_t *region;
+	char *name;
+
+	lib_printf(CONSOLE_BOLD "\n%-10s %-6s %-9s %-14s %-14s", "NAME", "REGION", "SUBREGION", "START", "END");
+	lib_printf("EN  XN  ACCESS P/U  TEX  S  C  B\n" CONSOLE_NORMAL);
+
+	for (i = 0; i < mpu_common.regCnt; i++) {
+		region = &mpu_common.hal.mpu.region[i];
+		if ((name = syspage_getMapName(mpu_common.hal.mpu.map[i])) == NULL)
+			name = "<invalid>";
+
+		mpu_regionPrint(name, region->rbar, region->rasr);
+	}
+
+	lib_printf("\nConfigured %d of %d MPU regions based on %d map definitions.\n",
+		mpu_common.regCnt, mpu_common.regMax, mpu_common.mapsCnt);
+}
+
+
+/* Reset structures and detect MPU type */
+int mpu_init(void)
+{
+	unsigned int i;
+
+	hal_memset(&mpu_common, 0, sizeof(mpu_common));
+
+	for (i = 0; i < MPU_REGION_NB; i++)
+		mpu_regionDisable(i);
+
+	mpu_common.hal.mpu.type = _imxrt_typeMPU();
+	mpu_common.regMax = (u8)(_imxrt_typeMPU() >> 8);
+
+	if (mpu_common.regMax != MPU_REGION_NB)
+		return -EFAULT;
+
+	mpu_common.flags |= mpu_fValid;
+
+	return 0;
+}
+
+
+/* Store MPU region setup into syspage_hal data */
+int mpu_syspageConfig(void)
+{
+	unsigned int i;
+
+	if ((mpu_common.flags & mpu_fValid) == 0) {
+		lib_printf("MPU context is not initialized.\n");
+		return -EFAULT;
+	}
+
+	if ((mpu_common.flags & mpu_fError) != 0) {
+		lib_printf("Could not continue due to MPU invalid map setup, please clear error flag.\n");
+		return -EFAULT;
+	}
+
+	/* Disable unused regions */
+	for (i = mpu_common.regCnt; i < mpu_common.regMax; i++) {
+		mpu_common.hal.mpu.map[i] = (u32)-1;
+		mpu_regionDisable(i);
+	}
+
+	mpu_common.hal.mpu.regions = mpu_common.regCnt;
+
+	syspage_setHalData(&mpu_common.hal);
+
+	mpu_common.flags |= mpu_fConfigured;
+
+	return 0;
+}
```

Comment:
Please don't use function from upper layers (hardware independent layers) in bottom layers (hardware dependent).

Reply:
Functionality removed.